### PR TITLE
Add basic ArrayPool cleaning tests

### PR DESCRIPTION
**Pull request by JeremyKuhne**
I'm checking basic behavior of both the thread locals and the stacks. Any additional test ideas are welcome.

### CODE REVIEW

File: src/System.Buffers/tests/ArrayPool/CollectionTests.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+namespace System.Buffers.ArrayPool.Tests
+{
+    [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+    public class CollectionTests : RemoteExecutorTestBase
+    {
+        private const string TrimSwitchName = "DOTNET_SYSTEM_BUFFERS_ARRAYPOOL_TRIMSHARED";
+
+        [Theory,
+            InlineData(true),
+            InlineData(false)]
```
Comment (stephentoub) [inline src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: OuterLoop for all of these.  I'm hesitant to have tests that take so long even in OuterLoop, but I know I differ there from @danmosemsft.

Comment (JeremyKuhne) [reply src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: This one I will for sure, I called out in the comments that I'm leaving it innerloop until I see it go CI green. The other one is actually fast (a few seconds).

**Comment by JeremyKuhne** [comment]
@stephentoub I pushed the event tests out of proc to avoid the cleanup events. I'll add a few specific ones for the new events.

**Comment by JeremyKuhne** [comment]
@dotnet-bot test Windows x64 Debug Build
@dotnet-bot test Windows x86 Release Build

System.Net & Cryptography failures

File: src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                    callback(e);
+                }, body);
+                return count;
+            }
+        }
+
+        protected static void RemoteInvokeWithTrimming(Action action, bool trim = false)
+        {
+            RemoteInvokeOptions options = new RemoteInvokeOptions();
+            options.StartInfo.UseShellExecute = trim;
```
Comment (jkotas) [inline src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs]: Should this rather be `UseShellExecute = false` ?

File: src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        protected static void RemoteInvokeWithTrimming(Func<string, int> method, bool trim = false, int timeout = FailWaitTimeoutMilliseconds)
+        {
+            RemoteInvokeOptions options = new RemoteInvokeOptions
+            {
+                TimeOut = timeout
+            };
+
+            options.StartInfo.UseShellExecute = trim;
```
Comment (jkotas) [inline src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs]: Same here.

File: src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Threading;
+
+namespace System.Buffers.ArrayPool.Tests
+{
+    public class ArrayPoolTest : RemoteExecutorTestBase
```
Comment (stephentoub) [inline src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs]: Nit: public abstract class ArrayPoolTestBase

File: src/System.Buffers/tests/ArrayPool/CollectionTests.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                    }
+                }
+
+                // Should only have found a new buffer if we're trimming
+                Assert.Equal(parsedTrim, foundNewBuffer);
+                return SuccessExitCode;
+            }, trim, 3 * 60 * 1000); // This test has to wait for the buffers to go stale (give it three minutes)
+        }
+
+        [OuterLoop("This test is risky as it allocates a ton of memory (can cause other tests to fail)")]
```
Comment (stephentoub) [inline src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: > can cause other tests to fail

Then we shouldn't have it even in outerloop.

Comment (jkotas) [reply src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: The tests that are stressing memory management in different ways (and take over the machine as side-effect) may be better suited for CoreCLR repo. I believe we have other GC tests in this category.

Comment (JeremyKuhne) [reply src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: Would this create issues in a stress environment too? The problem is running test assemblies in parallel.

File: src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Buffers/tests/ArrayPool/ArrayPoolTest.cs]: Sure

File: src/System.Buffers/tests/ArrayPool/CollectionTests.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: > Would this create issues in a stress environment too? The problem is running test assemblies in parallel.

We don't have a stress environment for these.  Enabling tests with the suggested attribute currently requires a dev to do it explicitly.  If/when we have better CI support for it, we'll need to address the fact that most such stress tests behave as if they own the box.

Comment (JeremyKuhne) [reply src/System.Buffers/tests/ArrayPool/CollectionTests.cs]: > My top-level point here is I don't want such tests running as part of outerloop jobs.

Yes, I have no problem with that. Just don't want to kick the ball forward to another currently running automated system. :)

**Comment by JeremyKuhne** [comment]
Addressed all of the feedback- if there is anything else I'll follow up

**Pull request merged**